### PR TITLE
[WIP] Added support for named params with regex path

### LINKED ISSUE #154: Support path params in RegExp path format
When declaring a route in the `Regexp` format, there's no way to include route params. I've checked the source code for `regexparam` and it seems to act quite weirdly with RegExp inputs (returning `false` as keys? non-sens...). That means the situation needs to be handled manually on our side by literally iterating through the RegExp string.

### PR DESCRIPTION
**Pull request by vigenere23**
fixes #154 (partially)

This PR is a proposal for adding support for name params (aka `:id` and etc.) while using `RegExp` paths. 

> Please note that this is a proposal for code example and is clearly not complete

### How it works

`regexparam` handles `RegExp` paths quite weirdly and returns `false` as `keys` instead of the actual wanted keys (which works well when `path` is a `string`). To make it work, we have to:

1. manually iterate over the string value of the `RegExp` path
2. find all occurences of named groups (starting with `?<` and ending with `>`)
3. adding the found named group in the path, minus the delimiters
4. replace that part of the path by adding an almost-wildcard pattern `[^\/]+?` just like it does for string path params (`:param`)
5. since this string comes from `RegExp.prototype.toString()`, it includes the `/` delimiters + the flag, which **should not** be present when constructing a new `RegExp`, so they are removed

### Not done in PR

- Tests
- Edge cases
- Optional params

### Disclaimer

Sorry if some weird Git behavior happens with this PR. This is my first fork merge, so I may not be familiar with the best pratices yet.

### CODE REVIEW

**Comment by vigenere23** [comment]
Sorry, had to reopen after branch renaming. Original PR: #153

File: Router.svelte [hunk 1]
```diff
@@ -269,9 +316,10 @@ class RouteItem {
             throw Error('Invalid value for "path" argument')
         }
 
-        const {pattern, keys} = regexparam(path)
+        // At this point path is valid and either a string of RegExp, so we can extract the pattern and path params
+        const {pattern, keys} = parsePath(path)
 
-        this.path = path
+        this.path = path // TODO remove unused
```
Comment (ItalyPaleAle) [inline Router.svelte:322]: This is used in the `detail` object for events (see line 501, or 453 before changes)

File: Router.svelte [hunk 2]
```diff
@@ -49,6 +49,53 @@ function getLocation() {
...
+    }
+    // Else path is a RegExp
+
+    const keys = []
+
+    // Named parameters in a RegExp path are represented with capture groups (`?<parameter_name>`).
+    // We need to extract the inside as key, and add an almost-wildcard
+    //  expression `([^\\/]+?)` that will match like a string path param.
+    let newPath = path.toString()
+        .replace(/[?]<.*>/g, (match) => {
```
Comment (ItalyPaleAle) [inline Router.svelte:75]: I would really prefer if this were an actual parser than some regular expression magic. Parsers built using a regexp are generally frowned upon because they might not be the most robust.

How about parsing with just a for loop iterating on every character, looking for the `(?` sequence?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: Router.svelte (hunk 2 above)
Comment (vigenere23) [reply Router.svelte:75]: Okok, I will do some search on how such a parser would ideally be implemented, I guess it should not be that hard ;p

File: Router.svelte (hunk 1 above)
Comment (vigenere23) [reply Router.svelte:322]: Woops, didn't see that in VSCode, will remove

File: Router.svelte (hunk 2 above)
Comment (vigenere23) [reply Router.svelte:75]: Would a BNF state-machine be something interesting in that case? I'm thinking about something like this :

<details>
<summary>Algorithm</summary>

```js
function parseRegexPath(path) {
    const keys = []
    const newPathCharacters = []
    const KEY_START_1 = '?'
    const KEY_START_2 = '<'
    const KEY_END = '>'
    const PATH_PARAM_REGEX = '([^\\/]+?)'

    let keyBuffer = []
    let state = STATE_NOT_KEY // some Enum, TS Union or set of constants

    path.toString().split('').forEach(character => {
        newPathCharacters.push(character)

        if (state === STATE_NOT_KEY && character === KEY_START_1) {
            state = STATE_MAYBE_KEY
        }
        else if (state === STATE_MAYBE_KEY && character === KEY_START_2) {
            state = STATE_KEY
        }
        else if (state === STATE_KEY) {
            if (character === KEY_END) {
                keys.push(keyBuffer.join(''))
                newPathCharacters.push(PATH_PARAM_REGEX)
                state = STATE_NOT_KEY
                keyBuffer = []
            } else {
                keyBuffer.push(character)
            }
        }
    })

    let newPath = newPathCharacters.join('')

    // ... some other final logic and return
}
```

</details>

**Comment by ItalyPaleAle** [comment]
Yes the state machine is what I was thinking about, although it can be simplified a bit.

However, as I started writing a parser, I stumbled upon something...

If I understand your PR correctly, it should transform named groups into `([^\\/]+?)`. However, shouldn't we just accept whatever RegExp the user wrote?

For example, assume a route used this regular expression: `/^\/hola\/(?<foo>.*)/`
We shouldn't try to rewrite `.*`, but rather just look into the matched groups by name and use them as that!

**Comment by vigenere23** [comment]
Yes, for that part i wasn't sure. Having to manually write `([^\/]+?)` after each and every path param is exhausting, but at the same time, isn't this the reason why you would use regex paths in the first place? So yeah, there's plenty of room for discussing about it, but you made me realize that in fact it may be better to let the user do it and not modify its input. 

For the parsing, I guess there's a faster/simpler way by using forward-looking `indexOf` or `find`, instead of literally looping on each character.

**Comment by ItalyPaleAle** [comment]
I'd prefer if we kept the regexp as-is. In fact, the reason for regular expressions was to allow more flexibility... Sometimes you might want `[^\/]+?`, but other times you will want to write your own custom match, such as `[0-9]{4,6}` or anything else! 

In that case, then we don't even need to parse the regular expression. Invoking `match` on a regexp that contains a named capture group should automatically return a dictionary of all named groups in the `.groups` property. See this example:
image

In this case, we might not even need to make any code change, as the `props` array should actually contain the `props.groups` property. It would just require some update to the documentation to explain it's possible.

I'll look into it tonight

**Comment by ItalyPaleAle** [comment]
Updated the docs with instructions on how to do this.

Thanks a lot for bringing this issue up and for proposing a fix with a PR!
